iree/base/internal/threading_test.cc

```diff
@@ -51,10 +51,6 @@ TEST(ThreadTest, Lifetime) {
                                     iree_allocator_system(), &thread));
   EXPECT_NE(0, iree_thread_id(thread));
 
-  // Drop the thread handle; should be safe as the thread should keep itself
```

Comment:
what was the issue here? as the comment notes part of the API is that threads should retain themselves to allow for fork-and-return - if they aren't that's a bug

Reply:
(see https://github.com/google/iree/blob/42b8134ea43c04d49b6c1146d0c2a466746e6642/iree/base/internal/threading_pthreads.c#L167-L170)

Reply:
Two separate TSan reports, one about a data race, one about a "thread leak" (didn't know that was a thing and you could call that one mere policy, but one has to admit that the fact that fixing that also fixed the data race, shows it wasn't a totally useless policy for TSan to have).

My interpretation of the data race itself was that the root issue was that by early-releasing the main thread's reference, we were causing thread destruction to occur on the other thread, racing with main thread at-exit code (`cxa_at_exit_wrapper`, `cxa_thread_atexit_impl`).

```
WARNING: ThreadSanitizer: data race (pid=975614)
  Read of size 8 at 0x7ff6cc5a0078 by main thread:
    #0 tracy::moodycamel::ConcurrentQueue<tracy::QueueItem, tracy::moodycamel::ConcurrentQueueDefaultTraits>::~ConcurrentQueue() /usr/local/google/home/benoitjacob/iree/third_party/tracy/client/tracy_concurrentqueue.h:489:13 (threading_test+0x4d2d4c)
    #1 cxa_at_exit_wrapper(void*) crtstuff.c (threading_test+0x30231f)

  Previous write of size 8 at 0x7ff6cc5a0078 by thread T2:
    #0 tracy::moodycamel::ProducerToken::~ProducerToken() /usr/local/google/home/benoitjacob/iree/third_party/tracy/client/tracy_concurrentqueue.h:354:20 (threading_test+0x4d2b44)
    #1 __call_tls_dtors stdlib/cxa_thread_atexit_impl.c:155:7 (libc.so.6+0x3f81e)

  As if synchronized via sleep:
    #0 nanosleep <null> (threading_test+0x2c32ff)
    #1 void std::this_thread::sleep_for<long, std::ratio<1l, 1000l> >(std::chrono::duration<long, std::ratio<1l, 1000l> > const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/this_thread_sleep.h:82:9 (threading_test+0x34451d)
    #2 (anonymous namespace)::ThreadTest_CreateSuspended_Test::TestBody() /usr/local/google/home/benoitjacob/iree/iree/base/internal/threading_test.cc:100:3 (threading_test+0x33c71d)
    #3 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2607:10 (threading_test+0x3c0607)
    #4 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2643:14 (threading_test+0x39192a)
    #5 testing::Test::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2682:5 (threading_test+0x35aa78)
    #6 testing::TestInfo::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2861:11 (threading_test+0x35bccf)
    #7 testing::TestSuite::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:3015:28 (threading_test+0x35c902)
    #8 testing::internal::UnitTestImpl::RunAllTests() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:5855:44 (threading_test+0x373a41)
    #9 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2607:10 (threading_test+0x3c6157)
    #10 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2643:14 (threading_test+0x395daa)
    #11 testing::UnitTest::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:5438:10 (threading_test+0x3732bd)
    #12 RUN_ALL_TESTS() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/include/gtest/gtest.h:2490:46 (threading_test+0x3496ad)
    #13 main /usr/local/google/home/benoitjacob/iree/iree/testing/gtest_main.cc:17:10 (threading_test+0x34961b)

  Thread T2 (tid=975617, finished) created by main thread at:
    #0 pthread_create <null> (threading_test+0x2c598d)
    #1 iree_thread_create /usr/local/google/home/benoitjacob/iree/iree/base/internal/threading_pthreads.c:180:10 (threading_test+0x347e63)
    #2 (anonymous namespace)::ThreadTest_Lifetime_Test::TestBody() /usr/local/google/home/benoitjacob/iree/iree/base/internal/threading_test.cc:50:3 (threading_test+0x33bb1d)
    #3 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2607:10 (threading_test+0x3c0607)
    #4 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2643:14 (threading_test+0x39192a)
    #5 testing::Test::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2682:5 (threading_test+0x35aa78)
    #6 testing::TestInfo::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2861:11 (threading_test+0x35bccf)
    #7 testing::TestSuite::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:3015:28 (threading_test+0x35c902)
    #8 testing::internal::UnitTestImpl::RunAllTests() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:5855:44 (threading_test+0x373a41)
    #9 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2607:10 (threading_test+0x3c6157)
    #10 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2643:14 (threading_test+0x395daa)
    #11 testing::UnitTest::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:5438:10 (threading_test+0x3732bd)
    #12 RUN_ALL_TESTS() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/include/gtest/gtest.h:2490:46 (threading_test+0x3496ad)
    #13 main /usr/local/google/home/benoitjacob/iree/iree/testing/gtest_main.cc:17:10 (threading_test+0x34961b)

SUMMARY: ThreadSanitizer: data race /usr/local/google/home/benoitjacob/iree/third_party/tracy/client/tracy_concurrentqueue.h:489:13 in tracy::moodycamel::ConcurrentQueue<tracy::QueueItem, tracy::moodycamel::ConcurrentQueueDefaultTraits>::~ConcurrentQueue()
==================
==================
WARNING: ThreadSanitizer: thread leak (pid=975614)
  Thread T2 (tid=975617, finished) created by main thread at:
    #0 pthread_create <null> (threading_test+0x2c598d)
    #1 iree_thread_create /usr/local/google/home/benoitjacob/iree/iree/base/internal/threading_pthreads.c:180:10 (threading_test+0x347e63)
    #2 (anonymous namespace)::ThreadTest_Lifetime_Test::TestBody() /usr/local/google/home/benoitjacob/iree/iree/base/internal/threading_test.cc:50:3 (threading_test+0x33bb1d)
    #3 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2607:10 (threading_test+0x3c0607)
    #4 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2643:14 (threading_test+0x39192a)
    #5 testing::Test::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2682:5 (threading_test+0x35aa78)
    #6 testing::TestInfo::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2861:11 (threading_test+0x35bccf)
    #7 testing::TestSuite::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:3015:28 (threading_test+0x35c902)
    #8 testing::internal::UnitTestImpl::RunAllTests() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:5855:44 (threading_test+0x373a41)
    #9 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2607:10 (threading_test+0x3c6157)
    #10 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:2643:14 (threading_test+0x395daa)
    #11 testing::UnitTest::Run() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/src/gtest.cc:5438:10 (threading_test+0x3732bd)
    #12 RUN_ALL_TESTS() /usr/local/google/home/benoitjacob/iree/third_party/googletest/googletest/include/gtest/gtest.h:2490:46 (threading_test+0x3496ad)
    #13 main /usr/local/google/home/benoitjacob/iree/iree/testing/gtest_main.cc:17:10 (threading_test+0x34961b)

SUMMARY: ThreadSanitizer: thread leak (/usr/local/google/home/benoitjacob/iree-build-linux/iree/base/internal/threading_test+0x2c598d) in pthread_create
```

Reply:
Note: is Tracy defeating some IREE assumption by injecting atexit handlers ? 😝

Reply:
@benvanik ping - rebased this, this is more timely to review now. This is just test-only issues, but now that we've made so much progress, there are only 9 remaining TSan failures and half of them are test-only, so this is a significant further step to green TSan.

I hope that the above comment with the 2 TSan reports answers your question. I agree that in the abstract, it's perfectly legitimate to say that what TSan flags as a "thread leak" is a normal idiom, even a simplification/optimization in some cases.

I still think that we should take this particular TSan report of "thread leak" seriously rather than silence it, for 2 reasons:

1. With the caveat that that was with Tracy instrumentation enable, there was another issue here, a data race also reported by TSan.
2. Fork-and-return is indeed very useful in practice and is part of a more general idea of "why make the end user wait for something to shutdown, if that shutdown is just boilerplate, not actually useful work". But while that's a good strategy in a release-build where the focus is on what's user-observable, it's still useful to keep a naive full-block shutdown in tests and debug builds and generally all programs built to catch bugs not to optimize user experience. Because without a normal shutdown, it's hard to tell if we're only leaking things intentionally or if we also have unintentional leaks (note TSan calls the present situation a "thread leak", which sounds pedantic at first, but a thread could own any other kind of resource so a "thread leak" is potentially a leak of anything).